### PR TITLE
ENH: support reductions for pyarrow temporal types

### PR DESCRIPTION
**Pull request by jbrockmendel**
We could plausibly disallow mean for date types, matching what we do for PeriodDtype.

Fixes 232 xfails, cuts test runtime by 88 seconds.

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -1083,6 +1090,24 @@ def pyarrow_meth(data, skip_nulls, **kwargs):
 
         if name in ["min", "max", "sum"] and pa.types.is_duration(pa_type):
             result = result.cast(pa_type)
+        if name in ["median", "mean"] and pa.types.is_temporal(pa_type):
+            result = result.cast(pa_type)
+        if name in ["std", "sem"] and pa.types.is_temporal(pa_type):
+            result = result.cast(pa.int64())
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:1096]: Is this int64 cast needed? I thought the result should already be an int type?

File: pandas/core/arrays/arrow/array.py [hunk 2]
```diff
@@ -1083,6 +1090,24 @@ def pyarrow_meth(data, skip_nulls, **kwargs):
...
+            if pa.types.is_duration(pa_type):
+                result = result.cast(pa_type)
+            elif pa.types.is_time(pa_type):
+                # TODO: with time types we should probably retain "unit",
+                #  but not clear how to get that since pa_type.unit raises
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:1101]: Is this on an older version of pyarrow?

```
In [6]: t = pa.time32("ms")

In [7]: t.unit
Out[7]: 'ms'

In [8]: pa.__version__
Out[8]: '10.0.1'

In [9]: t = pa.time64("ns")

In [10]: t.unit
Out[10]: 'ns'

In [11]: t = pa.time32("ms")

In [12]: t.unit
Out[12]: 'ms'
```

File: pandas/core/arrays/arrow/array.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1096]: im pretty sure its float64 at this point

File: pandas/core/arrays/arrow/array.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1101]: looks like it depends on exactly what we're checking.  Here pa_type is defined as `pa_type = self._data.type` which  `DataType(time32[ms])`, which does not have a .unit attribute in 10.0.1

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1101]: so how do I get the `Time32Type(time32[ms])` instead of the `DataType(time32[ms])` here?

Comment (jorisvandenbossche) [reply pandas/core/arrays/arrow/array.py:1101]: Yeah, this was only recently fixed (should be in pyarrow 11): https://github.com/apache/arrow/pull/14633

When getting the type from the array, it wasn't correctly boxed in the Time specific class, and so the unit attribute is not available. I am not directly sure of a workaround for older pyarrow versions, except for parsing that from the str repr of the type ...

Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1101]: so something like?

```
def get_unit_from_pyarrow_dtype(pa_dtype):
    if pa_lt_11:
        return str(pa_dtype).split("[")[-1][:-1]
    return pa_dtype.unit
```

Comment (jorisvandenbossche) [reply pandas/core/arrays/arrow/array.py:1101]: Yes, that should work I think

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**